pmd_wavefront/interfaces/genesis4.py

```diff
@@ -0,0 +1,117 @@
+                 3d complex dfl grid with shape (nx, ny, nz)
+   
+   field values in Genesis4 format
+        
+    """
+        
+    # Point to the real and imaginary h5 handles
+    E_complex = h5_meshes['electricField']
+    # Get attributes
+    attrs = dict(E_complex.attrs)
+    
+    gridsize, gridsize, slicespacing =  E_complex.attrs['gridSpacing']
+    refposition = E_complex.attrs['timeOffset']
+    wavelength = 12398.425 / E_complex.attrs['photonEnergy'] * 1.0e-10
+    
+    gridpoints, gridpoints, slicecount = np.array(E_complex['x'].shape)
+    params = dict(gridpoints=gridpoints, slicecount=slicecount, gridsize=gridsize, refposition=refposition, slicespacing=slicespacing, wavelength= wavelength)
+
+    factorGenesis = gridsize / np.sqrt(2.0 * Z0) #factor to genesis units
+    unitSI = E_complex['x'].attrs['unitSI'] # factor to convert to V/m
+    dflfactor = unitSI * factorGenesis # V/m -> Genesis
+    
+    assert ('x' in E_complex) or ('y' in E_complex), 'missing field components'
+    
+    # Get arrays
+    if 'x' in E_complex:
+        
+        dfl_x = dflfactor * E_complex['x'][:,:,:]
+        
+        if genesis4_filename:
+            write_dfl_to_genesis4_field(genesis4_filename+'_x.h5',  dflfactor * E_complex['x'][:,:,:], **params)
+        
+    else:
+        dfl_x = None
+        
+        
+    if 'y' in E_complex:
+        
+        dfl_y = dflfactor * E_complex['y'][:,:,:]
+        
+        if genesis4_filename:
+            write_dfl_to_genesis4_field(genesis4_filename+'_y.h5', dflfactor * E_complex['y'][:,:,:], **params)
+            
+    else:
+        dfl_y = None
+        
+    
+    return dfl_x, dfl_y
+
+
+def write_dfl_to_genesis4_field(filename, dfl, *, gridpoints, slicecount, gridsize, refposition, slicespacing, wavelength):
+    """
+    Write dfl array to an open H5 handle in genesis4 format.
+    
+    filename: str
+              Genesis 4 field file name
+    dfl: numpy.ndarray
+        3d complex dfl grid with shape (nx, ny, nz)
+    param: Genesis parameter dict. This routine extracts:
+        gridpoints (ncar in v2)
+        slicecount (nslice in v2)
+        gridsize (dgrid in v2)
+        refposition
+        slicespacing (zsep in v2)
+        wavelength (xlamds in v2)
+    to write the appropriate metadata.
+    
+    """
+    
+    h5 = h5py.File(filename, 'w')
+    
+    vars = [gridpoints, slicecount, gridsize, refposition, slicespacing, wavelength]
+    names = ["gridpoints", "slicecount", "gridsize", "refposition", "slicespacing", "wavelength"]
+    types = ['i4', 'i4', 'f8', 'f8', 'f8', 'f8']
+    
+    for var, name, type in zip(vars, names, types):
+        
+        dset = h5.create_dataset(name, (1,), dtype=type)
+        dset[:] = var
+    
+    dfl = dfl.reshape((gridpoints**2, slicecount))
+    
+    for i in range(0, slicecount):
+        ind = i + 1
+        g = h5.create_group('slice' + str(f'{ind:06}'))
+        g.create_dataset('field-real', (slicelength,), dtype='f8')
```

Comment:
You don't need these lines, you can write `g['field-real'] = np.real(dfl[:, I])` directly, it will make the dataset.